--- a/Makefile
+++ b/Makefile
@@ -25,3 +25,7 @@ ccflags-y += -Werror -Wall
 
 # clang
 ccflags-y += -Wno-frame-address -Wno-unused-but-set-parameter
+
+ifdef SFE_SUPPORT_IPV6
+ccflags-y += -DSFE_SUPPORT_IPV6
+endif
--- a/sfe_cm.c
+++ b/sfe_cm.c
@@ -16,8 +16,16 @@
  */
 
 #include <linux/module.h>
+#include <linux/version.h>
 #include <linux/sysfs.h>
 #include <linux/skbuff.h>
+#include <linux/inet.h>
+#include <linux/netfilter.h>
+
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0))
+int (*athrs_fast_nat_recv)(struct sk_buff *skb) __rcu __read_mostly;
+EXPORT_SYMBOL_GPL(athrs_fast_nat_recv);
+#endif
 #include <net/route.h>
 #include <net/ip6_route.h>
 #include <net/addrconf.h>
@@ -742,8 +750,13 @@ sfe_cm_ipv6_post_routing_hook(hooknum, o
 static int sfe_cm_conntrack_event(struct notifier_block *this,
 				  unsigned long events, void *ptr)
 #else
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0))
+static int sfe_cm_conntrack_event(unsigned int events, const struct nf_ct_event *item)
+#else
 static int sfe_cm_conntrack_event(unsigned int events, struct nf_ct_event *item)
 #endif
+
+#endif
 {
 #ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
 	struct nf_ct_event *item = ptr;
@@ -824,7 +837,11 @@ static struct notifier_block sfe_cm_conn
 };
 #else
 static struct nf_ct_event_notifier sfe_cm_conntrack_notifier = {
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0))
+	.ct_event = sfe_cm_conntrack_event,
+#else
 	.fcn = sfe_cm_conntrack_event,
+#endif
 };
 #endif
 #endif
@@ -1127,8 +1144,13 @@ static int __init sfe_cm_init(void)
 #ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
 	result = nf_conntrack_register_chain_notifier(&init_net, &sfe_cm_conntrack_notifier);
 #else
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0))
+	nf_conntrack_register_notifier(&init_net, &sfe_cm_conntrack_notifier);
+	result = 0;
+#else
 	result = nf_conntrack_register_notifier(&init_net, &sfe_cm_conntrack_notifier);
 #endif
+#endif
 	if (result < 0) {
 		DEBUG_ERROR("can't register nf notifier hook: %d\n", result);
 		goto exit4;
@@ -1215,9 +1237,13 @@ static void __exit sfe_cm_exit(void)
 #ifdef CONFIG_NF_CONNTRACK_CHAIN_EVENTS
 	nf_conntrack_unregister_chain_notifier(&init_net, &sfe_cm_conntrack_notifier);
 #else
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0))
+	nf_conntrack_unregister_notifier(&init_net);
+#else
 	nf_conntrack_unregister_notifier(&init_net, &sfe_cm_conntrack_notifier);
 #endif
 #endif
+#endif
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 13, 0))
 	nf_unregister_hooks(sfe_cm_ops_post_routing, ARRAY_SIZE(sfe_cm_ops_post_routing));
--- a/sfe_cm.h
+++ b/sfe_cm.h
@@ -150,7 +150,9 @@ struct sfe_connection_mark {
 /*
  * Expose the hook for the receive processing.
  */
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0))
 extern int (*athrs_fast_nat_recv)(struct sk_buff *skb);
+#endif
 
 /*
  * Expose what should be a static flag in the TCP connection tracker.
--- a/sfe_ipv4.c
+++ b/sfe_ipv4.c
@@ -16,6 +16,8 @@
  */
 
 #include <linux/module.h>
+
+MODULE_DESCRIPTION("Shortcut Forwarding Engine - IPv4 edition");
 #include <linux/sysfs.h>
 #include <linux/skbuff.h>
 #include <linux/icmp.h>
@@ -409,7 +411,7 @@ struct sfe_ipv4 {
 					/* Tail of the list of all connections */
 	unsigned int num_connections;	/* Number of connections */
 	struct timer_list timer;	/* Timer used for periodic sync ops */
-	sfe_sync_rule_callback_t __rcu sync_rule_callback;
+	void * __rcu sync_rule_callback;
 					/* Callback function registered by a connection manager for stats syncing */
 	struct sfe_ipv4_connection *conn_hash[SFE_IPV4_CONNECTION_HASH_SIZE];
 					/* Connection hash table */
@@ -1152,7 +1154,7 @@ static void sfe_ipv4_flush_sfe_ipv4_conn
 	rcu_read_lock();
 	spin_lock_bh(&si->lock);
 	si->connection_flushes++;
-	sync_rule_callback = rcu_dereference(si->sync_rule_callback);
+	sync_rule_callback = (sfe_sync_rule_callback_t)rcu_dereference(si->sync_rule_callback);
 	spin_unlock_bh(&si->lock);
 
 	if (sync_rule_callback) {
@@ -1474,7 +1476,9 @@ static int sfe_ipv4_recv_udp(struct sfe_
 	/*
 	 * Mark that this packet has been fast forwarded.
 	 */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 18, 0)
 	skb->fast_forwarded = 1;
+#endif
 
 	/*
 	 * Send the packet on its way.
@@ -2047,7 +2051,9 @@ static int sfe_ipv4_recv_tcp(struct sfe_
 	/*
 	 * Mark that this packet has been fast forwarded.
 	 */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 18, 0)
 	skb->fast_forwarded = 1;
+#endif
 
 	/*
 	 * Send the packet on its way.
@@ -2865,8 +2871,12 @@ static void sfe_ipv4_periodic_sync(struc
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0))
 	struct sfe_ipv4 *si = (struct sfe_ipv4 *)arg;
 #else
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0)
+	struct sfe_ipv4 *si = container_of(tl, struct sfe_ipv4, timer);
+#else
 	struct sfe_ipv4 *si = from_timer(si, tl, timer);
 #endif
+#endif
 	u64 now_jiffies;
 	int quota;
 	sfe_sync_rule_callback_t sync_rule_callback;
@@ -2874,7 +2884,7 @@ static void sfe_ipv4_periodic_sync(struc
 	now_jiffies = get_jiffies_64();
 
 	rcu_read_lock();
-	sync_rule_callback = rcu_dereference(si->sync_rule_callback);
+	sync_rule_callback = (sfe_sync_rule_callback_t)rcu_dereference(si->sync_rule_callback);
 	if (!sync_rule_callback) {
 		rcu_read_unlock();
 		goto done;
@@ -3585,7 +3595,11 @@ static void __exit sfe_ipv4_exit(void)
 	 */
 	sfe_ipv4_destroy_all_rules_for_dev(NULL);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0)
+	timer_delete_sync(&si->timer);
+#else
 	del_timer_sync(&si->timer);
+#endif
 
 	unregister_chrdev(si->debug_dev, "sfe_ipv4");
 
@@ -3613,6 +3627,6 @@ EXPORT_SYMBOL(sfe_register_flow_cookie_c
 EXPORT_SYMBOL(sfe_unregister_flow_cookie_cb);
 #endif
 
-MODULE_DESCRIPTION("Shortcut Forwarding Engine - IPv4 edition");
+
 MODULE_LICENSE("Dual BSD/GPL");
 
--- a/sfe_ipv6.c
+++ b/sfe_ipv6.c
@@ -16,6 +16,8 @@
  */
 
 #include <linux/module.h>
+
+MODULE_DESCRIPTION("Shortcut Forwarding Engine - IPv6 support");
 #include <linux/sysfs.h>
 #include <linux/skbuff.h>
 #include <linux/icmp.h>
@@ -436,7 +438,7 @@ struct sfe_ipv6 {
 					/* Tail of the list of all connections */
 	unsigned int num_connections;	/* Number of connections */
 	struct timer_list timer;	/* Timer used for periodic sync ops */
-	sfe_sync_rule_callback_t __rcu sync_rule_callback;
+	void * __rcu sync_rule_callback;
 					/* Callback function registered by a connection manager for stats syncing */
 	struct sfe_ipv6_connection *conn_hash[SFE_IPV6_CONNECTION_HASH_SIZE];
 					/* Connection hash table */
@@ -1212,7 +1214,7 @@ static void sfe_ipv6_flush_connection(st
 	rcu_read_lock();
 	spin_lock_bh(&si->lock);
 	si->connection_flushes++;
-	sync_rule_callback = rcu_dereference(si->sync_rule_callback);
+	sync_rule_callback = (sfe_sync_rule_callback_t)rcu_dereference(si->sync_rule_callback);
 	spin_unlock_bh(&si->lock);
 
 	if (sync_rule_callback) {
@@ -1513,7 +1515,9 @@ static int sfe_ipv6_recv_udp(struct sfe_
 	/*
 	 * Mark that this packet has been fast forwarded.
 	 */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 18, 0)
 	skb->fast_forwarded = 1;
+#endif
 
 	/*
 	 * Send the packet on its way.
@@ -2070,7 +2074,9 @@ static int sfe_ipv6_recv_tcp(struct sfe_
 	/*
 	 * Mark that this packet has been fast forwarded.
 	 */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 18, 0)
 	skb->fast_forwarded = 1;
+#endif
 
 	/*
 	 * Send the packet on its way.
@@ -2873,8 +2879,12 @@ static void sfe_ipv6_periodic_sync(struc
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0))
 	struct sfe_ipv6 *si = (struct sfe_ipv6 *)arg;
 #else
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0)
+	struct sfe_ipv6 *si = container_of(tl, struct sfe_ipv6, timer);
+#else
 	struct sfe_ipv6 *si = from_timer(si, tl, timer);
 #endif
+#endif
 	u64 now_jiffies;
 	int quota;
 	sfe_sync_rule_callback_t sync_rule_callback;
@@ -2882,7 +2892,7 @@ static void sfe_ipv6_periodic_sync(struc
 	now_jiffies = get_jiffies_64();
 
 	rcu_read_lock();
-	sync_rule_callback = rcu_dereference(si->sync_rule_callback);
+	sync_rule_callback = (sfe_sync_rule_callback_t)rcu_dereference(si->sync_rule_callback);
 	if (!sync_rule_callback) {
 		rcu_read_unlock();
 		goto done;
@@ -3593,7 +3603,11 @@ static void __exit sfe_ipv6_exit(void)
 	 */
 	sfe_ipv6_destroy_all_rules_for_dev(NULL);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0)
+	timer_delete_sync(&si->timer);
+#else
 	del_timer_sync(&si->timer);
+#endif
 
 	unregister_chrdev(si->debug_dev, "sfe_ipv6");
 
@@ -3620,6 +3634,6 @@ EXPORT_SYMBOL(sfe_ipv6_register_flow_coo
 EXPORT_SYMBOL(sfe_ipv6_unregister_flow_cookie_cb);
 #endif
 
-MODULE_DESCRIPTION("Shortcut Forwarding Engine - IPv6 support");
+
 MODULE_LICENSE("Dual BSD/GPL");
 
